### PR TITLE
Avoid reindexing on a single level rollback

### PR DESCRIPTION
**Pull request by droserasprout**

### CODE REVIEW

**Comment by droserasprout** [comment]
[Single level rollback](https://trello.com/c/jA3gbVNb/130-single-level-rollback)

File: src/dipdup/datasources/tzkt/datasource.py [hunk 1]
```diff
@@ -629,7 +664,7 @@ def convert_operation(cls, operation_json: Dict[str, Any]) -> OperationData:
             type=operation_json['type'],
             id=operation_json['id'],
             level=operation_json['level'],
-            timestamp=operation_json['timestamp'],
+            timestamp=datetime.fromisoformat(operation_json['timestamp'][:-1]).replace(tzinfo=timezone.utc),
```
Comment (m-kus) [inline src/dipdup/datasources/tzkt/datasource.py:667]: Helper?

File: src/dipdup/index.py [hunk 2]
```diff
@@ -139,16 +168,34 @@ async def _synchronize(self, last_level: int) -> None:
...
+        if self._rollback_level:
+            if state.level != self._rollback_level:
+                raise RuntimeError(f'Rolling back to level {self._rollback_level}, state level {state.level}')
+            if level != self._rollback_level:
+                raise RuntimeError(f'Rolling back to level {self._rollback_level}, got operations of level {level}')
+
+            self._logger.info('Rolling back to previous level, verifying processed operations')
+            expected_hashes = set([opg.hash for opg in self._rollback_operation_groups])
+            received_hashes = set([op.hash for op in operations])
+            if expected_hashes != received_hashes:
```
Comment (m-kus) [inline src/dipdup/index.py:185]: should be `received_hashes.intersection(expected_hashes) == expected_hashes`

File: src/dipdup/models.py [hunk 3]
```diff
@@ -48,6 +51,15 @@ class Meta:
         abstract = True
 
 
+class StateOperationGroup(Model):
```
Comment (m-kus) [inline src/dipdup/models.py:54]: Switching to in-memory opg hashes cache (as discussed)

**Review by m-kus — APPROVED** [review]

**Pull request merged**